src/scheduler/queue.c

```diff
@@ -0,0 +1,177 @@
+ *
+ * Commercial License Information:
+ *
+ * For a copy of the commercial license terms and conditions,
+ * go to: (http://www.pbspro.com/UserArea/agreement.html)
+ * or contact the Altair Legal Department.
+ *
+ * Altair’s dual-license business model allows companies, individuals, and
+ * organizations to create proprietary derivative works of PBS Pro and
+ * distribute them - whether embedded or bundled with other software -
+ * under a commercial license agreement.
+ *
+ * Use of Altair’s trademarks, including but not limited to "PBS™",
+ * "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+ * trademark licensing policies.
+ *
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <errno.h>
+
+#include "constant.h"
+#include "log.h"
+#include "queue.h"
+
+
+/**
+ * @brief	Constructor for the data structure 'queue'
+ *
+ * @param	void
+ *
+ * @return queue *
+ * @retval a newly allocated queue object
+ * @retval NULL for malloc error
+ */
+ds_queue *
+new_ds_queue(void)
+{
+	ds_queue *ret_obj = NULL;
+
+	ret_obj = malloc(sizeof(ds_queue));
+	if (ret_obj == NULL) {
+		log_err(errno, __func__, MEM_ERR_MSG);
+		return NULL;
+	}
+
+	ret_obj->min_size = QUEUE_DS_MIN_SIZE;
+	ret_obj->content_arr = NULL;
+	ret_obj->front = 0;
+	ret_obj->rear = 0;
+	ret_obj->q_size = 0;
+
+	return ret_obj;
+}
+
+/**
+ * @brief	Destructor for a queue object
+ *
+ * @param[in]	obj - a queue object to deallocate
+ *
+ * @return void
+ */
+void
+free_ds_queue(ds_queue *obj)
+{
+	free(obj->content_arr);
+	free(obj);
+}
+
+
+/**
+ * @brief	Enqueue an object into the queue
+ *
+ * @param[in]	queue - the queue to enqueue the object in
+ * @param[in]	obj - the object to enqueue
+ *
+ * @return int
+ * @retval 1 for Success
+ * @retval 0 for Failure
+ */
+int
+ds_enqueue(ds_queue *queue, void *obj)
+{
+	long curr_rear;
+	long curr_qsize;
+
+	if (queue == NULL || obj == NULL)
+		return 0;
+
+	curr_rear = queue->rear;
+	curr_qsize = queue->q_size;
+
+	if (curr_rear >= curr_qsize) {
+		long new_qsize;
+		void **realloc_ptr = NULL;
+
+		/* Need to resize the queue */
+		if (curr_qsize == 0) /* First enqueue operation */
+			new_qsize = queue->min_size;
+		else
+			new_qsize = 2 * curr_qsize;
+
+		realloc_ptr = realloc(queue->content_arr, new_qsize * sizeof(void *));
+		if (realloc_ptr == NULL) {
+			log_err(errno, __func__, MEM_ERR_MSG);
+			return 0;
+		}
+
+		queue->content_arr = realloc_ptr;
+		queue->q_size = new_qsize;
+	}
+
+	queue->content_arr[curr_rear] = obj;
+	queue->rear = curr_rear + 1;
+
+	return 1;
+}
+
+/**
+ * @brief	Dequeue an object from the queue
+ *
+ * @param[in]	queue - the queue to use
+ *
+ * @return void *
+ * @retval the first item in queue
+ * @retval NULL for error/empty queue
+ */
+void *
+ds_dequeue(ds_queue *queue)
+{
+	if (queue == NULL)
+		return NULL;
+
+	if (queue->front == queue->rear) /* queue is empty */
+		return NULL;
+
+	return queue->content_arr[queue->front++];
```

Comment:
Does this mean that objects will be there in the queue even after they are dequeued? I think we should memcpy and realloc after dequeuing some number of objects.

Reply:
Thanks for catching this, right now the queue can grow forever! But I think it might be better  if we reset the front and rear pointers when the queue gets empty. This will allow better use of it once the queue reaches a big enough size and will prevent the queue from growing forever. So, it'll get used like a buffer. What do you think?

Reply:
I like this Idea